keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,388 @@
+* the new feature would need to be enabled by default to always apply the bug fix for new releases.
+* it changes the API which is not desirable in patch version releases.
+
+The proper way of handling cases like `ProgressNotify` is:
+1. the feature should be gated by the feature gate from the beginning.
+1. the feature should be disabled by default until it is widely tested in practice. 
+1. when the bus is found, the feature should ideally be at a lifecycle in which it is disabled by default. If not, the admin should disable it by the `--feature-gates` flag.
+1. when the client upgrades etcd to the patch version with the fix, the admin could enable it by the `--feature-gates` flag.
+
+#### Could the lifecycle of a feature change in patch versions?
+
+Kubernetes have a minor release every 3 months, while the cadence of etcd minor releases is much less frequent. The question is: do we have to wait for years before graduating a new feature?
+
+We think we should still stick to the common practice of not changing the lifecycle of a feature in patch versions. Because:
+* changing the lifecycle of a feature is an API change. According to the [etcd Operations Guide](https://etcd.io/docs/v3.5/op-guide/versioning/), only new minor versions may add additional features to the API.
+* bugs in etcd could be hard to detect, and the reliability and robustness of etcd is more important than speed. A long history of testing through practical adoption a new feature is beneficial.
+
+With the feature gate in place, we could consider increasing the etcd release cadence because it would be easier to add new features and less risky to release new features.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New APIs
+
+New `--feature-gates` flag would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.
+
+New grpc and http endpoints would be added to query if a feature is enabled for the server or cluster. New `etcdctl` will also be added.
+
+For grpc, new rpc type would be added to the `Maintenance` service.
+
+```proto
+service Maintenance {
+  ...
+  rpc FeatureGateStatus(FeatureGateStatusRequest) returns (FeatureGateStatusResponse) {
+    option (google.api.http) = {
+      post: "/v3/maintenance/featuregate"
+      body: "*"
+    };
+  }
+  ...
+}
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+
+message FeatureGateStatusRequest {
+  // if true, query if the features are enabled for the single server.
+  // otherwise, query if the features are enabled for the cluster.
+  bool isServerFeature = 1;
```

Comment:
For simplicity, suggest not to add the concept of "cluster level feature gate". If inconsistency across all nodes in the cluster may cause problem, then let's clearly clarify it in the document, and print warning/error in log if possible. If the inconsistency may cause critical issue (i.e. data inconsistency), then we can even fail the etcd's startup (but I do not see such case so far).

Also this is a generic issue, not limited to only feature gate. etcd has lots of flags, the inconsistencies across all nodes for each flag may cause issue or confusion.
```suggestion
```

Reply:
inconsistencies is inevitable in upgrade/downgrade scenarios. For example a feature can be alpha (disabled by default) in one version, and beta (enabled by default) in the next. I think we have to have a clear way of handling the inconsistencies. Just warning/error messages is not enough.

Reply:
- Name `clusterLevelFeature` makes more sense then `isServerFeature`. But see bigger concern below
- With respect to the inconsistence, 
  - As mentioned above, inconsistence is a generic issue when upgrading/updating a cluster, not only limited to the feature gate. For example, when we change the value of a flag (it can be any flag, including the feature gate) during upgrading or updating a cluster, there is a small window of inconsistence. So if we really want to resolve such issue, it's out of the scope of feature gate, we should have a more generic solution.
  - But I don't see any major or critical issue for such inconsistence. So we need to have a list of such issues firstly, and clearly clarify each issue, then discuss the solution. It can be discussed separately.

Reply:
feature gate is a potential solution to inconsistencies. 
* if a flag is available in both versions, it is probably ok to assume they are consistently set during upgrade/downgrade.
* if a flag is available only in one version, ideally it should be gated behind a feature gate, and feature gate would gracefully handle the case when it is set in one version but not in the other. 

Let's imagine a hypothetical case:
1. a new write type is introduced in V_n+1, and some K8s function depends on this new feature.
2. K8s asks an etcd server (which happens to be V_n+1) if the feature is enabled, and gets confirmation, and starts sending the new write requests.
3. the new write requests would be sent to etcd leader (which happens to be V_n), and it does not recognize this new write type. 

However short this intermediate mixed version state is, the workload could potentially break the whole cluster. Right now, we have to introduce a new raft type in V_n before using it in V_n+1 to prevent such failures. But if we can handle cluster feature enablement, we could remove the step of introducing a new raft type in V_n. 
I think there is great benefit in speeding up the dev process with this cluster level feature gate.

Reply:
Another example is the ProgressNotify, which [KEP-2340](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md) depends on. So K8s performance can be greatly impacted by whether this feature is available or not. If there is no cluster level feature information, depending on which api server is the master, or which etcd server it talks to, ProgressNotify could flip flop, and making the performance inconsistent. There is no guarantee etcd upgrade/downgrade could be done in x seconds, I think we should make our best to make the intermediate state predictable. 

cc @serathius

Reply:
Cluster level flags are not outside the scope of the KEP. Let's get back to the discussion when the new proposal will be ready for review.